tokio/tests/io_mem_stream.rs

```diff
@@ -62,20 +64,47 @@ async fn disconnect() {
     t2.await.unwrap();
 }
 
+#[tokio::test]
+async fn disconnect_reader() {
+    let (a, mut b) = duplex(2);
+
+    let t1 = tokio::spawn(async move {
+        // this will block, as not all data fits into duplex
+        b.write_all(b"ping").await.unwrap_err();
+    });
+
+    let t2 = tokio::spawn(async move {
+        // here we drop the reader side, and we expect the writer in the other
+        // task to exit with an error
+        let _moved = a;
+    });
+
+    // wait for drop first
+    t2.await.unwrap();
+    // then try and join writer task
+    t1.await.unwrap();
```

Comment:
These comment seem to imply that the order is important, but changing the order cannot change the test's behavior.

Reply:
isn't the await necessary to ensure that t2 is scheduled to run before t1? or is that not necessary in tests/wrong in general? otherwise this would just test the existing behaviour of error on starting a new write on a closed stream

Reply:
When you `tokio::spawn` something, it starts running immediately. The returned `JoinHandle` is only useful for waiting for the task — Tokio takes care of running it regardless of whether you use the `JoinHandle`.

Reply:
To be clear, the test does currently not guarantee any ordering between `t1` and `t2` whatsoever. They may run in any order or even at the same time.

Reply:
Hm, I see, thanks. Would it be worth it then to guarantee the order somehow? In my testing this would always reliably always trigger the bug, but both cases are supposed to behave the same on the writer side anyway. If we decide to leave it, I'd just remove the comments.

Reply:
If you want `t2` to run first, then `t1`, then just don't spawn them and put them in that order.

Reply:
But the write_all blocks, so I need to await it in a spawned task? Anyway, for me it reproduces the issue just fine and shows the fix working, so I've just removed the comments for now.

Reply:
I think it's fine like this as well.